content/en/blog/_posts/2024-03-19-go-workspaces.md

```diff
@@ -0,0 +1,176 @@
+didn't love it. They designed a new way of stitching multiple modules together
+into a new "workspace" concept. Once enrolled in a workspace, Go tools had
+enough information to work in any directory structure and across modules,
+without GOPATH or symlinks or other dirty tricks.
+
+When I first saw this proposal I knew that this was the way out. This was how
+to break the logjam. If workspaces was the technical solution, then I would
+put in the work to make it happen.
+
+## The work
+
+Adopting workspaces was deceptively easy. I very quickly had the codebase
+compiling and running tests with workspaces enabled. I set out to purge the
+repository of anything GOPATH related. That's when I hit the first real bump -
+the code-generation tools.
+
+We had about a dozen tools, totalling several thousand lines of code. All of
+them were built using an internal framework called
+[gengo](https://github.com/kubernetes/gengo), which was built on Go's own
+parsing libraries. There were 2 main problems:
+
+  1) Those parsing libraries didn't understand modules or workspaces.
+  2) GOPATH allowed us to pretend that Go "package paths" and directories on
+     disk were interchangeable in trivial ways. They are not.
+
+Switching to a [modules- and workspaces-aware parsing](https://pkg.go.dev/golang.org/x/tools@v0.19.0/go/packages)
+library was the first step. Then I had to make a long series of changes to
+each of the code-generation tools. Critically, I had to find a way to do it
+that was possible for some other person to review! After some conversations
+with likely reviewers, we agreed that a series of small commits would be
+easiest to review.
+
+Sadly, these were not mechanical changes. I had to dig into each tool to
+figure out where they were processing disk paths versus where they were
+processing package names, and where those were being conflated. I made
+extensive use of the [delve](https://github.com/go-delve/delve) debugger, which
+I just can't say enough good things about.
+
+One unfortunate result of this work was that I had to break compatibility. The
+gengo library simply did not have enough information to process packages
+outside of GOPATH. After discussion with gengo and Kubernetes maintainers, we
+agreed to make [gengo/v2](https://github.com/kubernetes/gengo/tree/master/v2).
+I also used this as an opportunity to clean up some of the gengo APIs and the
+tools' CLIs to be more understandable and not conflate packages and
+directories, for example you can't just string-join directory names and
+assume the result is a valid package name.
+
+Once I had the code-generation tools converted, I shifted attention to the
+dozens of scripts in the `hack` directory. One by one I had to run them, debug,
+and fix failures. Some of them needed minor changes and some needed to be
+rewritten.
+
+Along the way we hit cases that Go did not support, like workspace vendoring,
+which they then [implemented](https://github.com/golang/go/issues/60056)
```

Comment:
I think one of the most interesting aspects of this whole project is the collaboration between K8s and Go communities. Any details to share on how this collaboration happened? (Presumably 'they' are real people interested in the developer experience and they implemented this based on some input from you - is that right? If so, it would be great to share if possible without becoming a huge tangent.

Reply:
Added a bit more words and called out thanks to the Go team for hearing us (in the last section).  PTAL or suggest more detailed idea?

Reply:
LGTM.